Presentation layer structure document (#222)
software-platform/monorepo: metrics/web/docs/02_presentation_layer_architecture.md

2. High-level widgets

`High-level widgets` are used to actually display the data for a user. Commonly, these widgets use `low-level widgets` as a building components to create the required view. Also, these widgets should accept the `view model`. See [Widget structure organization](03_widget_structure_organization.md) document to get more information about `high-level` widgets.

So, to sum up: 

1. Page - the widget that properly combines the high-level widgets. Represents the web page or screen of the application. These widgets lay under the `module_name/presentation/pages` folder.
2. High-level widgets - the widgets that may, or may not, consist of low-level widgets and properly configure them. It should accept the `view model` on input. High-level widgets can obtain any other params like `ThemeStrategy` to make it more testable.
3. Low-level widgets - the configurable widgets that simply displays the data. They should accept the Dart's native types, like `int`, `String`, `Point`, `bool` (can accept types from any UI packages if necessary). They should be highly reusable. Low-level widgets should be placed under the `common/presentation` package.

> > Explain the way of using the `strings` in the widgets and where to place them.

Reviewer: ```suggestion
### Strings
```

To create a `high-level` widget, we commonly should use the `low-level` widgets in the implementation and create a `view model` for this widget.

To create a `low-level` widget, we should implement the widget that will be highly-configurable and reusable in different parts of the application.

## Widget structure
> Create the class diagram explaining the structure of widgets.

Let's consider a class diagram explaining the structure of widgets in the Metrics Web Application: 

![Widget Structure Class Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/software-platform/monorepo/presentation_layer_structure_document/metrics/web/docs/diagrams/widget_structure_class_diagram.puml)

> Explain the package structure of the presentation layer.

The package structure is also an important part of the Metrics Web Application presentation layer. Every presentation unit has a similar package structure. Let's consider the example of the `cool_module`'s package structure.

> * cool_module/
>    * data/...
>    * domain/...
>    * presentation/

> >       * view_model/

Reviewer: ```suggestion
>       * view_models/
```
Should we change it like that? If yes, then we should probably update other documents as well...

> What will be impacted by the project?

- The implementation of presentation layer components of the Metrics Web Application is impacted.

> - Since a developer should consider creating low-level widgets and then use it for creating the desired module-specific one, hence the process of delivering a presentation layer component is impacted. 

Reviewer: ```suggestion
```

# Testing
> How will the project be tested?

The presentation layer will be unit-tested and integration-tested using the core [flutter_test](https://api.flutter.dev/flutter/flutter_test/flutter_test-library.html) package and the [flutter_driver](https://api.flutter.dev/flutter/flutter_driver/flutter_driver-library.html) package respectively.

# Alternatives Considered
> Summarize alternative designs (pros & cons)

- Not document the presentation layer of the Metrics Web Application:
    - Cons: 
        - As the application grows, the future development of new modules and maintaining the old ones may be tricky without any document describing the presentation layer.

>         - As there is no explanation in differences between top-level and low-level widgets in the context of the Metrics Web Application, newcomers will be required to investigate this by themselves. This point makes the Metrics Web Application entry threshold quite hight.

Reviewer: ```suggestion
        - With no written explanation in differences between top-level and low-level widgets in the context of the Metrics Web Application, it will be longer for the new team members to become productive.
```
